src/MacroExpander.js

```diff
@@ -25,18 +28,87 @@ MacroExpander.prototype.nextToken = function() {
+        var tok;
         var expansion = this.macros[name];
         if (typeof expansion === "string") {
+            var numArgs = 0;
+            if (expansion.indexOf("#") !== -1) {
+                var stripped = expansion.replace(/##/g, "");
+                while (stripped.indexOf("#" + (numArgs + 1)) !== -1) {
+                    ++numArgs;
+                }
+            }
             var bodyLexer = new Lexer(expansion);
             expansion = [];
-            var tok = bodyLexer.lex();
+            tok = bodyLexer.lex();
             while (tok.text !== "EOF") {
                 expansion.push(tok);
                 tok = bodyLexer.lex();
             }
             expansion.reverse(); // to fit in with stack using push and pop
+            expansion.numArgs = numArgs;
             this.macros[name] = expansion;
         }
+        if (expansion.numArgs) {
+            var args = [];
+            var i;
+            // obtain arguments, either single token or balanced {…} group
+            for (i = 0; i < expansion.numArgs; ++i) {
+                var startOfArg = this.get(true);
+                if (startOfArg.text === "{") {
+                    var arg = [];
+                    var depth = 1;
+                    while (depth !== 0) {
+                        tok = this.get(false);
+                        arg.push(tok);
+                        if (tok.text === "{") {
+                            ++depth;
+                        } else if (tok.text === "}") {
+                            --depth;
+                        } else if (tok.text === "EOF") {
+                            throw new ParseError(
+                                "End of input in macro argument",
+                                startOfArg);
+                        }
+                    }
+                    arg.pop(); // remove last }
+                    arg.reverse(); // like above, to fit in with stack order
+                    args[i] = arg;
+                } else if (startOfArg.text === "EOF") {
+                    throw new ParseError(
+                        "End of input expecting macro argument", topToken);
+                } else {
+                    args[i] = [startOfArg];
+                }
+            }
+            // paste arguments in place of the placeholders
+            expansion = expansion.slice(); // make a shallow copy
+            for (i = expansion.length - 1; i >= 0; --i) {
+                tok = expansion[i];
+                if (tok.text === "#") {
+                    if (i === 0) {
+                        throw new ParseError(
+                            "Incomplete placeholder at end of macro body",
+                            tok);
+                    }
+                    tok = expansion[--i]; // next token on stack
+                    if (tok.text === "#") { // ## → #
+                        expansion.splice(i + 1, 1); // drop first #
+                    } else if (/^[1-9]$/.test(tok.text)) {
```

Comment:
The whole MacroExpander is working kind of backwards. It uses an array as a stack, with tokens added by lexing (one at a time) or macro expansion (may be many), and tokens removed by consumption by the parser (one at a time). The canonical way of using an array as a stack is using its `push` and `pop` methods, which modify the tail end of the array. So if subsequent `pop` calls should return the tokens in order, the stack itself has to have them in reverse order, with higher indices corresponding to eralier tokens. That's the reason I have two calls to `reverse` in there by now.

If this is causing too much confusion, we might add bigger comments than those we have, but I guess those would still get lost in the diffs you see on GitHub. Or we could make the stack work forward, and hope that a `shift` is fast enough. Or we could avoid removing elements, and instead maintain a pointer indicating the current position. Do you think any of these approaches would be better than what we have now?

I guess that although the difference between `push`/`pop` and `shift`/`unshift` [appears to be considerable](https://jsperf.com/push-pop-vs-unshift-shift/5), it's probably still negligible compared to the other processing we do, so I'm willing to reformulate all of this in terms of these if you want me to.

See https://github.com/Khan/KaTeX/pull/493#discussion-diff-68474688 for previous discussion on this.